src/mpi/datatype/typerep/src/typerep_yaksa_pack.c

```diff
@@ -152,8 +152,12 @@ static int typerep_do_pack(const void *inbuf, MPI_Aint incount, MPI_Datatype dat
     }
 
     MPL_pointer_attr_t inattr, outattr;
-    MPIR_GPU_query_pointer_attr(inbuf_ptr, &inattr);
-    MPIR_GPU_query_pointer_attr(outbuf, &outattr);
+
+    /* only query the attributes for contiguous memory */
+    if (is_contig) {
+        MPIR_GPU_query_pointer_attr(inbuf_ptr, &inattr);
+        MPIR_GPU_query_pointer_attr(outbuf, &outattr);
+    }
```

Comment:
In yaksa pr, the hack is limited to data buffer `MPI_BOTTOM`, can we do the same in MPICH? I am not sure routing all noncontiguous datatypes to the slow path will always be preferred.

Reply:
Agree. I changed the code to only avoid attribute caching in case of absolute addressing.